api/v1alpha1/externaldns_types.go

```diff
@@ -18,41 +18,412 @@ package v1alpha1
+
+	// Credentials is a reference to a secret containing
+	// the relevant credentials for the given provider, to be
+	// used by an ExternalDNS controller for creating, updating,
+	// and viewing DNS records.
+	//
+	// +kubebuilder:validation:Required
+	// +required
+	Credentials *corev1.LocalObjectReference `json:"credentials"`
+}
+
+// ExternalDNSProviderUnion describes optional fields for an ExternalDNS
+// provider that should be captured.
+// +union
+type ExternalDNSProviderUnion struct {
+	// Type describes which DNS provider
+	// ExternalDNS should publish records to.
+	// The following DNS providers are supported:
+	//
+	//  * AWS
+	//  * GCP
+	//  * Azure
+	//  * BlueCat
+	//  * Infoblox
+	//
+	// +kubebuilder:validation:Required
+	// +unionDiscriminator
+	// +required
+	Type ExternalDNSProviderType `json:"type"`
+}
+
+// +kubebuilder:validation:Enum=AWS;GCP;Azure;BlueCat;Infoblox
+type ExternalDNSProviderType string
+
+const (
+	ProviderTypeAWS      ExternalDNSProviderType = "AWS"
+	ProviderTypeGCP      ExternalDNSProviderType = "GCP"
+	ProviderTypeAzure    ExternalDNSProviderType = "Azure"
+	ProviderTypeBlueCat  ExternalDNSProviderType = "BlueCat"
+	ProviderTypeInfoblox ExternalDNSProviderType = "Infoblox"
+	// More providers will ultimately be added in the future.
+)
+
+// ExternalDNSSource describes which Source resource
+// the ExternalDNS should create DNS records for.
+type ExternalDNSSource struct {
+	ExternalDNSSourceUnion `json:",inline"`
+
+	// HostnameAnnotationPolicy specifies whether or not ExternalDNS
+	// should ignore the "external-dns.alpha.kubernetes.io/hostname"
+	// annotation, which overrides DNS hostnames on a given source resource.
+	//
+	// The following values are accepted:
+	//
+	//  "Ignore": Ignore any hostname annotation overrides.
+	//  "Allow": Allow all hostname annotation overrides.
+	//
+	// The default behavior of the ExternalDNS is "Ignore".
+	//
+	// +kubebuilder:validation:Required
+	// +kubebuilder:default:=Ignore
+	// +required
+	HostnameAnnotationPolicy *HostnameAnnotationPolicy `json:"hostnameAnnotation"`
+}
+
+// ExternalDNSSourceUnion describes optional fields for an ExternalDNS source that should
+// be captured.
+// +union
+type ExternalDNSSourceUnion struct {
+	// Type specifies an ExternalDNS source resource
+	// to create DNS records for.
+	//
+	// +unionDiscriminator
+	// +kubebuilder:validation:Required
+	// +required
+	Type ExternalDNSSourceType `json:"type"`
+
+	// AnnotationFilter describes an annotation filter
+	// used to filter which source instance resources
+	// ExternalDNS publishes records for.
+	// The annotation filter uses label selector semantics
+	// against source resource annotations.
+	//
+	// +optional
+	AnnotationFilter map[string]string `json:"annotationFilter,omitempty"`
+
+	// Namespace instructs ExternalDNS to only acknowledge
+	// source resource instances in a specific namespace.
+	//
+	// +optional
+	Namespace string `json:"namespace,omitempty"`
```

Comment:
This looks severely limiting.  I would expect namespace selection to use label selection where the individual namespace name could be used as a selector if desired.

Reply:
Alas, this is a shortcoming of ExternalDNS, which can only observe all namespaces, or one namespace per instance (and on top of that, only one source resource per instance).
https://github.com/kubernetes-sigs/external-dns/issues/1454
We could support a namespace label selector here, but the fan out wouldn't be obvious.